tests/std/tests/P2374R4_views_cartesian_product/test.cpp

```diff
@@ -0,0 +1,1032 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <forward_list>
+#include <list>
+#include <ranges>
+#include <span>
+#include <string_view>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include <range_algorithm_support.hpp>
+
+using namespace std;
+using ranges::bidirectional_range, ranges::random_access_range, ranges::sized_range, ranges::common_range,
+    ranges::iterator_t;
+
+template <bool B, class T>
+using maybe_const = conditional_t<B, const T, T>;
+
+// Helper concepts from [range.cartesian.view]
+template <bool IsConst, class First, class... Rest>
+concept CartesianProductIsRandomAccess = (random_access_range<maybe_const<IsConst, First>> && ...

```

Comment:
I believe the rationale is to double-check the implementation against an independent translation of the Standard's exposition-only concepts into code, so that if the implementation is modified in the future, we could notice any discrepancies. If we directly use `_Ugly` internals in test code (which we try to avoid doing, although not 100% strictly), we lose any chance of noticing such behavioral differences.

Reply:
Yeah, I don't have a huge problem with this, but it's not a super high value test, esp since the concepts are ugly so their behavior doesn't matter as long as everything else works.

Reply:
Thanks. These pretty concepts are just used to check the observable properties that are controlled by the ugly concepts. I agree that the test coverage value is relatively low (these fundamentals are unlikely to break), but it's consistent with our usual view testing where we try to inspect all of the major observable properties, and even low value coverage ends up being valuable in large quantities (as we have seen where later compiler or library features end up finding surprise interactions in simple pieces of test code that had always previously passed).